psrqpy/search.py

```diff
@@ -18,586 +18,2043 @@
+                PML[idx] = sc.pm_l_cosb.value
+
+                self.update(PMB, name='PMB')
+                self.update(PML, name='PML')
+
+    def derived_binary(self):
+        """
+        Calculate derived binary system parameters.
+        """
+
+        MASS_PSR = 1.35  # canonical pulsar mass (solar masses)
 
-        if loadfromfile:
-            self.load(loadfromfile)
+        # derive mass function
+        reqpars = ['A1', 'PB']
+        if np.all([p in self.columns for p in reqpars]):
+            A1 = self.catalogue['A1'].values.copy()*c.value
+            PB = self.catalogue['PB'].values.copy()*86400.
+
+            idx = np.isfinite(A1) & np.isfinite(PB)
+
+            MASSFN = np.full(self.catalogue_len, np.nan)
+            MASSFN[idx] = (4.*np.pi**2/GM_sun.value)*A1[idx]**3/(PB[idx]**2)
+
+            self.update(MASSFN, name='MASSFN')
+
+            # derive minimum, median and 90% UL for mass
+            MINMASS = np.full(self.catalogue_len, np.nan)
+            MEDMASS = np.full(self.catalogue_len, np.nan)
+            UPRMASS = np.full(self.catalogue_len, np.nan)
+            from scipy.optimize import newton
+
+            def solfunc(m2, sini, mf, m1):
+                return (m1 + m2)**2 - (m2*sini)**3/mf
+
+            SINI_MIN = 1.0  # inclination for minimum mass
+            SINI_MED = 0.866025403  # inclination of 60deg for median mass
+            SINI_90 = 0.438371146   # inclination for 90% UL mass
+            for i, mf in enumerate(MASSFN):
+                if ~np.isfinite(mf):
+                    continue
+
+                try:
+                    MINMASS[i] = newton(solfunc, MASS_PSR,
+                                        args=(SINI_MIN, mf, MASS_PSR),
+                                        maxiter=1000)
+                except RuntimeError:
+                    MINMASS[i] = np.nan
+                try:
+                    MEDMASS[i] = newton(solfunc, MASS_PSR,
+                                        args=(SINI_MED, mf, MASS_PSR),
+                                        maxiter=1000)
+                except RuntimeError:
+                    MEDMASS[i] = np.nan
+                try:
+                    UPRMASS[i] = newton(solfunc, MASS_PSR,
+                                        args=(SINI_90, mf, MASS_PSR),
+                                        maxiter=1000)
+                except RuntimeError:
+                    UPRMASS[i] = np.nan
+
+            self.update(MINMASS, name='MINMASS')
+            self.update(MEDMASS, name='MEDMASS')
+            self.update(UPRMASS, name='UPRMASS')
+
+        # derive eccentricity from EPS1 and EPS2
+        reqpars = ['EPS1', 'EPS2', 'ECC', 'OM']
+        if np.all([p in self.columns for p in reqpars]):
+            EPS1 = self.catalogue['EPS1'].values.copy()
+            EPS2 = self.catalogue['EPS2'].values.copy()
+            ECCnew = np.full(self.catalogue_len, np.nan)
+            OMnew = np.full(self.catalogue_len, np.nan)
```

Comment:
This section appears to assume that `ECC` and `OM` never exist and must always be derived from `EPS1` and `EPS2`, but that's not the case, as far as I know. Occasionally, people will publish `OM` and `ECC` values, so this calculation should go both ways.

Reply:
Yes, I will change things, so that `EPS1` and `EPS2` also get calculated.

Reply:
I've added calculations for `EPS1`, `EPS2` and their respective errors now.